src/mrinufft/io/pulseq.py

```diff
@@ -0,0 +1,255 @@
+        The system options for the Pulseq sequence. Default is `pp.Opts.default`.
+
+    Notes
+    -----
+    The  Sequence cycle can be summarized as follows:
+
+    1. RF pulse
+    2. Delay to sync TE
+    3. Gradients plays: The gradients consist in a prewind to the first point
+    of the trajectory, the trajectory itself, and a rewind to the edge of
+    k-space.
+    3bis. The ADC is opened on the trajectory points (ignoring the prewind and
+    rewinds parts)
+    4. Gradient spoilers
+    5. Delay to sync the next TR
+
+    Returns
+    -------
+    pp.Sequence
+        A Pulseq sequence object with the specified arbitrary gradient waveform.
+    """
+    TR = TR / 1000.0  # convert to seconds
+    TE = TE / 1000.0  # convert to seconds
+    seq = pp.Sequence(system=system)
+
+    seq.set_definition("FOV", fov)
+    seq.set_definition("ImgSize", img_size)
+    seq.set_definition("TR", TR)
+    seq.set_definition("TE", TE)
+    seq.set_definition("FA", FA)
+
+    if rf_pulse is None and FA is not None:
+        rf_pulse = pp.make_block_pulse(flip_angle=FA, system=system, use="excitation")
+    elif rf_pulse is not None and FA is not None:
+        raise ValueError(
+            "Cannot specify both `rf_pulse` and `FA`. Please provide only one."
+        )
+    elif rf_pulse is None and FA is None:
+        raise ValueError("Either `rf_pulse` or `FA` must be provided.")
+    if not isinstance(rf_pulse, SimpleNamespace):
+        raise TypeError(
+            "The `rf_pulse` parameter must be a SimpleNamespace object, "
+            "as returned by `pypulseq.make_block_pulse` or `pypulseq.make_arbitrary_rf`"
+        )
+
+    full_grads, skip_start, skip_end = prepare_trajectory_for_seq(
+        trajectory,
+        fov=fov,
+        img_size=img_size,
+        raster_time=system.grad_raster_time,
+        gamma=system.gamma,
+        gmax=system.max_grad,
+        smax=system.max_slew,
+        pregrad="prephase",
+        postgrad="slowdown_to_edge",
+    )
+    full_grads = full_grads * system.gamma  # convert to rad/s/m
+    traj_length = full_grads.shape[1] - skip_start - skip_end + 1
+    shot_duration = system.grad_raster_time * traj_length  # in seconds
+    adc = pp.make_adc(
+        num_samples=traj_length * osf,
+        system=system,
+        duration=shot_duration,
+        delay=skip_start * system.grad_raster_time,
+    )
+
+    # Add a spoiler gradient that move to twice the edge of k-space in the x direction.
+    spoiler = pp.make_trapezoid(
+        channel="x", area=2 * img_size[0] / fov[0], system=system
+    )
+
+    delay_before_grad = pp.make_delay(
+        TE
+        - pp.calc_rf_center(rf_pulse)[0]
+        - rf_pulse.delay
+        - TE_pos * shot_duration
+        - skip_start * system.grad_raster_time
+    )
+    delay_end_TR = pp.make_delay(
+        TR
+        - pp.calc_duration(rf_pulse)
+        - delay_before_grad.delay
+        - full_grads.shape[1] * system.grad_raster_time
+    )
```

Comment:
You might need to check for min_TE and min_TR. Can you add a util function to get it?
Basically for a fixed trajectory and spoil levels, give the minimum TE and TR possible.

Reply:
this is handled by the `check_timing` function of pulseq (there is check for delay that can't be negative for instance)